src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs

```diff
@@ -553,23 +550,30 @@ private void RegisterEventsToWorkspace(ReferenceCountedDisposable<State> state,
 
                 state.Target.Workspace.DocumentActiveContextChanged += OnActiveContextChanged;
                 state.Target.Workspace.Services.GetRequiredService<IWorkspaceStatusService>().StatusChanged += OnWorkspaceStatusChanged;
+                state.Target.Workspace.WorkspaceChanged += OnWorkspaceChanged;
             }
 
-            private void OnActiveContextChanged(object sender, DocumentActiveContextChangedEventArgs e)
-            {
-                // REVIEW: it would be nice for changed event to pass in both old and new document.
-                OnSuggestedActionsChanged(e.Solution.Workspace, e.NewActiveContextDocumentId, e.Solution.WorkspaceVersion);
-            }
-
-            private void OnDiagnosticsUpdated(object sender, DiagnosticsUpdatedArgs e)
+            private void OnWorkspaceChanged(object? sender, WorkspaceChangeEventArgs e)
             {
-                // document removed case. no reason to raise event
-                if (e.Solution == null)
+                switch (e.Kind)
                 {
-                    return;
+                    // Only care about document changes here, the call to OnSuggestedActionsChange
+                    // will filter and make sure the event only gets sent if its relevant to this buffer.
+                    case WorkspaceChangeKind.DocumentAdded:
+                    case WorkspaceChangeKind.DocumentRemoved:
+                    case WorkspaceChangeKind.DocumentReloaded:
+                    case WorkspaceChangeKind.DocumentChanged:
+                        OnSuggestedActionsChanged(e.NewSolution.Workspace, e.DocumentId, e.NewSolution.WorkspaceVersion);
+                        break;
+                    default:
+                        break;
                 }
+            }
 
-                OnSuggestedActionsChanged(e.Workspace, e.DocumentId, e.Solution.WorkspaceVersion);
+            private void OnActiveContextChanged(object sender, DocumentActiveContextChangedEventArgs e)
+            {
+                // REVIEW: it would be nice for changed event to pass in both old and new document.
+                OnSuggestedActionsChanged(e.Solution.Workspace, e.NewActiveContextDocumentId, e.Solution.WorkspaceVersion);
             }
```

Comment:
it's generally unclear to me why we need this code at all :)

Reply:
this is for telling VS to refresh the lightbulb when things have changed (e.g. file changes, context changes) which could change the lightbulb icon we show